### PR TITLE
Selinux support

### LINKED ISSUE #22: SELinux support
We need to setup SELinux policy for openSUSE/SLE to be able to provide SELinux capabilities for the underlying kubernetes engine. See https://github.com/rancher/rke2-selinux as an example

### PR DESCRIPTION
**Pull request by davidcassany**
This is a work in progress PR about setting and enabling SELInux in cOS-Toolkit

Fixes #22

### CODE REVIEW

**Comment by davidcassany** [comment]
So far I did few findings regarding the SELinux story around #22

* I could not manage to apply labels in a build process, dockers complaints of not being permitted. Fails with:
  ```
  /sbin/setfiles: Could not set context for /:  Operation not permitted
  ```
  The build is executed as root and, if I am not mistaken, the build runs with privileged capabilities. I don't know what further permissions/capabilities can be granted.

* As far as I could see xattr are mostly not preserved into container layers, buildah only preserves `security.capabilities` and few more, but not SELinux labels (see https://github.com/containers/buildah/issues/2127). Similar story in Moby, there is a PR to support it raised concerns and it is still opened for more than a year now https://github.com/moby/moby/pull/40855

* As a result of the above the obvious solution is to relabel the filesystem once the system boots. However this fails on the concept of the immutable root. We mount the root-tree as read only, so it miserably fails to relabel most parts of the root-tree. In addition, using the  build of [selinux-policies](https://build.opensuse.org/package/show/security:SELinux/selinux-policy?expand=0) for Leap 15.3 from OBS causes a reboot after relabeling, which in turn causes a reboot loop as the relabel gets lost at reboot.

* Another alternative could be applying the relabeling at install/upgrade face, IMHO this is the way to go. We need to be careful this does not invalidate the root-tree checksum so mtree reports the same result with and without SELinux labels.

* Since there are no selinux policies for SLE I am wondering what are we supposed to deliver here, adapting refpolicy to be functional with cOS seams to be a huge task. It is just frightening to see that fedora selinux polices are already a fork of it (due to the amount of differences) and the selinux-policies in TW are heavily patched over the fedora ones (and they are not submitted to SLE)... If the time arrives we require SELinux I am confident it is possible to get it given some suited policies, how to get those policies is the though part and probably something big (I am not a SELinux expert I can't asses the complexity of it)

* SELinux requires again some further changes on the kernel command line parameters (security=selinux selinux=1). With the current setup this not easy to achieve via upgrades as the grub configuration files do not belong to the image itself. IMHO this is design error, to build an image nothing no further bits from what the image contains should be required, meaning that in `iso/cOS.yaml` nothing beyond `system/cos` should be needed, I would include grub and its configuration in `system/cos` and then copy what ever is needed into the boot partition at installation time. I believe this way would be easier to design upgrades that bootloader sensitive. Related to #109

The current PR is mere informative about how I executed the tests, I don't think it is worth to merge it as it is now, it does not work and it will not work unless policies are refined (I'd say substantially refined).

**Comment by davidcassany** [comment]
Finally I managed to install the TW selinux packages, apply selinux labels at install time and then boot. The boot procedure now is on permissive mode and after boot one can see labels are applied on the rootfs (e.g. `ls -Z /`), if one provides the  `enforce=1` as a kernel parameter then SELinux is enfornced and the system boots with the only failures of `cos-setup` services, I guess this is a good sign as there are no specific policies for the boot time configurations executed by  `cos-setup`. Even the system boots, since `cos-setup` miserably failed to execute any stage the system is not configured as requested, including the root password, which makes the system unusable.

**Comment by davidcassany** [comment]
Apparently [go-mtree](https://github.com/vbatts/go-mtree#format) optionally supports parsing xattrs, it is an option. So I was wondering if it wouldn't be possible to actually include the go-mtree result of labeled files into the pushed image, so we could eventually dump the image to the filesystem then apply the labels and after that run go-mtree and compare with the pre-stored result that was included into image metadata. No labels processing on boot time, just at install/upgrade time and with a proper sanity check before rebooting. This way we could eventually avoid having to push a filesystem image to the registry.

**Comment by davidcassany** [comment]
#147 seams to be a valid procedure to verify the images however this still has a problem, which is that we need to somehow provide the mtree manifest as part of the image. Applying SELinux labels to the rootfs at build time even if it is just to create the mtree manifest might be tricky since my attempts to relabel inside the container all failed. Gonna dig it a little more, if this turns not to be possible this might be  a headache for building the image since a two step process would be required: one to build the image and another one to unpack, apply labels and repack without running the container itself.

**Comment by davidcassany** [comment]
@mudler @Itxaka @kkaempf  This PR is the first step of having SELinux support in cOS. This installs the policies and sets the labels at upgrade/install time. Currently SELinux is started on permissive mode to ensure SELinux does not block cloud-init operations. Proper policies should be applied to ensure all the changes we do on the fly aren't blocked.

Proper policies is something TBC and probably not something we should spend time on, since we are not experts on it.

**Review by mudler — APPROVED** [review]

File: packages/selinux-policies/build.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+requires:
+- name: "base"
+  category: "distro"
+  version: ">=0"
+
+steps:
+{{ if .Values.distribution }}
```
[the remark this answers is not in the document]
Comment (davidcassany) [reply packages/selinux-policies/build.yaml:7]: yes it should be possible to apply custom policies. Something similar to what is documented [here](https://debian-handbook.info/browse/en-US/stable/sect.selinux.html)  for the `Adapting the Rules` section. I can give it a try to see if I can adapt or provide some basic module even if it is just to meant to be an example.

[the remark this answers is not in the document]
Comment (mudler) [reply packages/selinux-policies/build.yaml:7]: @davidcassany thx :+1: even a link to that doc would be enough for the moment

File: values/opensuse.yaml [hunk 2]
```diff
@@ -35,5 +35,6 @@ packages: >-
     haveged
     tar
     rsync
+    timezone
```
Comment (Itxaka) [inline values/opensuse.yaml:38]: why is this needed? Does selinux require this?

**Review by Itxaka — APPROVED** [review]

File: values/opensuse.yaml (hunk 2 above)
Comment (davidcassany) [reply values/opensuse.yaml:38]: Good point, not related to SELinux, this is to fix an old glitch we have since the very beginning. Should have noted that on the commit message.

This line here https://github.com/rancher-sandbox/cOS-toolkit/blob/213b48663dc7b96e4cd77446a3e1a77dcf7c97df/packages/cloud-config/oem/01_defaults.yaml#L15

works because `UTC` is always the default, but `systemctl-firstboot` utility expects to find whatever we add here into `/usr/share/zoneinfo/*` and all those timezone definitions are provided by `timezone`. This is like locale setting, defaults to POSIX if you have none installed, but if you choose something else its required to have `glibc-locale` in the system which provides the `/usr/lib/locale/*` files.

File: packages/selinux-policies/build.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidcassany) [reply packages/selinux-policies/build.yaml:7]: All right, I managed to include a custom dummy cOS SELinux Targeted policy module. Now it is possible to boot cOS with enforced SELinux. The cOS module is the outcome of booting and upgrading on permissive mode without it, then I parsed audit logs with `audit2allow` tool which translates SELinux denials into some rules to avoid each specific denial. Obviously this is a bad approach, since it does not add any sort of specific hardening, it just relaxes the defaults probably in some not accurate way. I did not analyze if the result makes sense at all. I just checked I could boot with SELinux on `enforcing` mode.

**Comment by davidcassany** [comment]
Merging since all required tests are passing and pull_request suite passed all clean. Only one test of the push suite failed...
